File: app/src/main/java/sixue/naivereader/provider/QbxsProvider.java
package sixue.naivereader.provider;

import android.content.Context;
import android.util.Log;

import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import sixue.naivereader.Utils;
import sixue.naivereader.data.Book;
import sixue.naivereader.data.BookKind;
import sixue.naivereader.data.Chapter;
import sixue.naivereader.data.Source;
import sixue.naivereader.helper.OnlineHelper;

public class QbxsProvider extends NetProvider {
    private static final String TAG = QbxsProvider.class.getSimpleName();

    @Override
    public String getProviderId() {
        return "www.quanben.io";
    }

    @Override
    public String getProviderName() {
        return "全本小说";
    }

    @Override
    public List<Book> search(String s, Context context) {
        List<Book> list = new ArrayList<>();
        try {
            String key = URLEncoder.encode(s, "UTF-8");
            String url = "http://www.quanben.io/index.php?c=book&a=search&keywords=" + key;
            Connection.Response response = Jsoup.connect(url).followRedirects(true).timeout(5000).execute();
            if (!url.equals(response.url().toString())) {
                Log.d(TAG, response.url().toString());
                // cant reach.
            } else {
                Document doc = response.parse();

                for (Element item : doc.body().select(".list2")) {
                    Element titleEle = item.select("h3 a[itemprop=url]").first();
                    Element authorEle = item.select("p span[itemprop=author]").first();
                    Element coverEle = item.select("img").first();

                    String title = titleEle.text();
                    String id = titleEle.text();
                    String author = authorEle.text();
                    String para = titleEle.attr("href");
                    String coverUrl = coverEle.attr("src");

                    Book book = new Book();
                    book.setId(id);
                    book.setTitle(title);
                    book.setAuthor(author);
                    book.setKind(BookKind.Online);

                    Source source = new Source();
                    source.setId(getProviderId());
                    source.setPara(para);
                    book.getSources().add(source);

                    book.setSiteId(source.getId());
                    book.setSitePara(source.getPara());

                    OnlineHelper helper = (OnlineHelper) book.buildHelper();
                    helper.downloadCover(context, coverUrl);

                    list.add(book);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return list;
    }

    @Override
    public List<Chapter> downloadContent(Book book, String bookSavePath) {
        String para = book.getSitePara();
        String contentUrl = String.format("http://www.quanben.io%slist.html", para);
        List<Chapter> content = new ArrayList<>();
        try {
            Document doc = Jsoup.connect(contentUrl).timeout(5000).get();
            Elements elements = doc.body().select(".list3");
            for (Element ch : Jsoup.parse(elements.toString()).select("li[itemprop=itemListElement]")) {
                String title = ch.select("a").text();
                String url = ch.select("a").attr("href").trim();
                if (url.length() == 0) {
                    continue;
                }

                String id = url.substring(url.lastIndexOf("/"));

                final Chapter chapter = new Chapter();
                chapter.setId(url);
                chapter.setTitle(title);
                String chapterSavePath = calcChapterSavePath(chapter, bookSavePath);
                chapter.setSavePath(chapterSavePath);

                content.add(chapter);
            }
        } catch (IOException e) {
            Log.e(TAG, "downloadContent ERROR: " + contentUrl);
        }
        return content;
    }

    @Override
    public void downloadChapter(Book book, Chapter chapter) {

        String chapterUrl = getChapterUrl(book, chapter);
        try {
            String myJs = null;
            Document doc = Jsoup.connect(chapterUrl).timeout(5000).get();
            for (Element jsEle : doc.body().select("script[type=text/javascript]")) {
                String js = jsEle.html();
                if (js.startsWith("setTimeout(\"ajax_post(")) {
                    myJs = js;
                    break;
                }
            }
            if (myJs == null) {
                return;
            }

            myJs = myJs.substring("setTimeout(\"ajax_post(".length());
            myJs = myJs.substring(0, myJs.indexOf("\""));

            String[] paras = myJs.replace("'", "").split(",");
            if (paras.length < 10) {
                return;
            }

            String ajaxUrl = String.format("http://www.quanben.io/index.php?c=%s&a=%s", paras[0], paras[1]);
            String tm = "" + new Date().getTime();
            Document doc2 = Jsoup.connect(ajaxUrl)
                    .referrer(chapterUrl)
                    .data(paras[2], paras[3]) // pinyin, duorushenyuan
                    .data(paras[4], paras[5]) // id, 1
                    .data(paras[6], paras[7]) // "sky", "bce7cae095b38a070b33746694953743"
                    .data(paras[8], paras[9]) // "t", "1562330872"
                    .data("_type", "ajax")
                    .data("rndval", tm)
                    .timeout(5000).post();
            String text = doc2.html();
            String plainText = Utils.clearHtmlTag(text, new String[] {})
                    .replace("<p>", "").replace("</p>", "\n");
            Utils.writeText(plainText, chapter.getSavePath());
        } catch (IOException e) {
            Log.e(TAG, "downloadChapter ERROR: " + chapterUrl);
        }
    }

    @Override
    public String getChapterUrl(Book book, Chapter chapter) {
        return "http://www.quanben.io" + chapter.getId();
    }

    private String calcChapterSavePath(Chapter chapter, String bookSavePath) {
        return bookSavePath + "/" + chapter.getId().replace("/", "_").replace(".html", ".txt");
    }
}

File: app/src/main/java/sixue/naivereader/provider/QbxsProvider.kt
package sixue.naivereader.provider

import android.content.Context
import android.util.Log
import org.jsoup.Jsoup
import sixue.naivereader.Utils
import sixue.naivereader.data.Book
import sixue.naivereader.data.BookKind
import sixue.naivereader.data.Chapter
import sixue.naivereader.data.Source
import java.io.IOException
import java.net.URLEncoder

class QbxsProvider : NetProvider() {
    override val providerId: String
        get() = "www.quanben.io"
    override val providerName: String
        get() = "全本小说"

    override fun search(s: String, context: Context): List<Book> {
        val list: MutableList<Book> = ArrayList()
        try {
            val key = URLEncoder.encode(s, "UTF-8")
            val url = "https://www.quanben-xiaoshuo.com/?c=book&a=search&keyword=$key"
            val response = Jsoup.connect(url).followRedirects(true).timeout(5000).execute()
            if (url != response.url().toString()) {
                Log.d(TAG, response.url().toString())
                // cant reach.
            } else {
                val doc = response.parse()
                for (item in doc.body().select(".book")) {
                    val titleEle = item.select("h1[itemprop=name] a").first()!!
                    val authorEle = item.select("p span[itemprop=author]").first()!!
                    val title = titleEle.text()
                    val id = titleEle.text()
                    val author = authorEle.text()
                    val para = titleEle.attr("href")
                    val book = Book(
                        id = id,
                        title = title,
                        author = author,
                        kind = BookKind.Online,
                        localPath = "",
                    )
                    val source = Source(
                        id = providerId,
                        para = para
                    )
                    book.sources += source
                    book.siteId = source.id
                    book.sitePara = source.para
                    list.add(book)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return list
    }

    override fun parseBookUrl(s: String, context: Context): String? {
        return null
    }

    override fun loadBookUrl(s: String, context: Context): Book? {
        return null
    }

    override fun downloadContent(book: Book, bookSavePath: String): List<Chapter> {
        val para = book.sitePara!!
        val contentUrl = String.format("https://www.quanben-xiaoshuo.com%sxiaoshuo.html", para)
        val content: MutableList<Chapter> = ArrayList()
        try {
            val doc = Jsoup.connect(contentUrl).timeout(5000).get()
            val elements = doc.body().select(".list")
            for (ch in Jsoup.parse(elements.toString()).select("li[itemprop=itemListElement]")) {
                val title = ch.select("a").text()
                val url = ch.select("a").attr("href").trim { it <= ' ' }
                if (url.isEmpty()) {
                    continue
                }
//                val id = url.substring(url.lastIndexOf("/"))
                val chapter = Chapter(
                    id = url,
                    title = title,
                )
                val chapterSavePath = calcChapterSavePath(chapter, bookSavePath)
                chapter.savePath = chapterSavePath
                content.add(chapter)
            }
        } catch (e: IOException) {
            Log.e(TAG, "downloadContent ERROR: $contentUrl")
        }
        return content
    }

    override fun downloadChapter(book: Book, chapter: Chapter) {
        val chapterUrl = getChapterUrl(book, chapter)
        try {
            val doc = Jsoup.connect(chapterUrl).timeout(5000).get()
            val articleBody = doc.body().select("#articlebody").first()!!
            val text = articleBody.html()
            val plainText = Utils.clearHtmlTag(text, arrayOf())
                    .replace("<p>", "").replace("</p>", "\n")
            Utils.writeText(plainText, chapter.savePath)
        } catch (e: IOException) {
            Log.e(TAG, "downloadChapter ERROR: $chapterUrl")
        }
    }

    override fun getChapterUrl(book: Book, chapter: Chapter): String {
        return "https://www.quanben-xiaoshuo.com" + chapter.id
    }

    private fun calcChapterSavePath(chapter: Chapter, bookSavePath: String): String {
        return bookSavePath + "/" + chapter.id.replace("/", "_").replace(".html", ".txt")
    }

    companion object {
        private val TAG = QbxsProvider::class.java.simpleName
    }
}